File: src/main/java/hiiragi283/material/api/capability/item/HiiragiItemHandlerWrapper.java
package hiiragi283.material.api.capability.item;

import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraftforge.common.util.Constants;
import net.minecraftforge.common.util.INBTSerializable;
import net.minecraftforge.items.IItemHandlerModifiable;
import org.jetbrains.annotations.NotNull;

import javax.annotation.ParametersAreNonnullByDefault;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.List;

@ParametersAreNonnullByDefault
public class HiiragiItemHandlerWrapper implements IItemHandlerModifiable, INBTSerializable<NBTTagCompound> {

    private final List<AbstractMap.SimpleEntry<HiiragiItemHandler, Integer>> pairs = new ArrayList<>();

    public HiiragiItemHandlerWrapper(HiiragiItemHandler... itemHandlers) {
        //Wrapperのスロット -> IItemHandlerとそれに対応したスロットのEntry
        for (HiiragiItemHandler handler : itemHandlers) {
            for (int slot = 0; slot < handler.getSlots(); slot++) {
                pairs.add(new AbstractMap.SimpleEntry<>(handler, slot));
            }
        }
    }

    public AbstractMap.SimpleEntry<HiiragiItemHandler, Integer> getHandler(int slot) {
        validateSlotIndex(slot);
        return pairs.get(slot);
    }

    protected void validateSlotIndex(int slot) {
        if (slot < 0 || slot >= pairs.size())
            throw new RuntimeException("Slot " + slot + " not in valid range - [0," + pairs.size() + ")");
    }

    //    IItemHandlerModifiable    //

    @Override
    public int getSlots() {
        return pairs.size();
    }

    @NotNull
    @Override
    public ItemStack getStackInSlot(int slot) {
        validateSlotIndex(slot);
        return getHandler(slot).getKey().getStackInSlot(getHandler(slot).getValue());
    }

    @NotNull
    @Override
    public ItemStack insertItem(int slot, @NotNull ItemStack stack, boolean simulate) {
        validateSlotIndex(slot);
        if (canInsert(slot)) {
            return getHandler(slot).getKey().insertItem(getHandler(slot).getValue(), stack, simulate);
        } else return stack;
    }

    public boolean canInsert(int slot) {
        validateSlotIndex(slot);
        return getHandler(slot).getKey().getIOType().canInsert;
    }

    @NotNull
    @Override
    public ItemStack extractItem(int slot, int amount, boolean simulate) {
        validateSlotIndex(slot);
        if (canExtract(slot)) {
            return getHandler(slot).getKey().extractItem(getHandler(slot).getValue(), amount, simulate);
        } else return ItemStack.EMPTY;
    }

    public boolean canExtract(int slot) {
        validateSlotIndex(slot);
        return getHandler(slot).getKey().getIOType().canExtract;
    }

    @Override
    public int getSlotLimit(int slot) {
        validateSlotIndex(slot);
        return getHandler(slot).getKey().getSlotLimit(getHandler(slot).getValue());
    }

    @Override
    public void setStackInSlot(int slot, ItemStack stack) {
        validateSlotIndex(slot);
        getHandler(slot).getKey().setStackInSlot(getHandler(slot).getValue(), stack);
    }

    //    INBTSerializable    //

    @Override
    public NBTTagCompound serializeNBT() {
        var tagList = new NBTTagList();
        for (int slot = 0; slot < pairs.size(); slot++) {
            var pair = getHandler(slot);
            ItemStack stack = pair.getKey().getStackInSlot(pair.getValue());
            if (!stack.isEmpty()) {
                var tag = new NBTTagCompound();
                tag.setInteger("slot", slot);
                stack.writeToNBT(tag);
                tagList.appendTag(tag);
            }
        }
        var ret = new NBTTagCompound();
        ret.setTag("Items", tagList);
        return ret;
    }

    @Override
    public void deserializeNBT(NBTTagCompound nbt) {
        var tagList = nbt.getTagList("Items", Constants.NBT.TAG_COMPOUND);
        for (int index = 0; index < tagList.tagCount(); index++) {
            NBTTagCompound tag = tagList.getCompoundTagAt(index);
            int slot = tag.getInteger("Slot");
            if (slot >= 0 && slot < pairs.size()) {
                var pair = getHandler(slot);
                pair.getKey().setStackInSlot(pair.getValue(), new ItemStack(tag));
            }
        }
    }

}

File: src/main/kotlin/hiiragi283/material/api/capability/item/HiiragiItemHandlerWrapper.kt
package hiiragi283.material.api.capability.item

import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.nbt.NBTTagList
import net.minecraftforge.common.util.Constants
import net.minecraftforge.common.util.INBTSerializable
import net.minecraftforge.items.IItemHandlerModifiable

/**
 * RagiItemHandlerとスロットの番号を紐づけたPairの一覧を作ることで，複数のRagiItemHandlerをまとめて処理できる
 * Thanks to SkyTheory!
 */

class HiiragiItemHandlerWrapper(vararg itemHandlers: HiiragiItemHandler)
    : IItemHandlerModifiable, INBTSerializable<NBTTagCompound> {

    private val pairs: MutableList<Pair<HiiragiItemHandler, Int>> = mutableListOf()

    init {
        for (itemHandler in itemHandlers) {
            for (slot in 0 until itemHandler.slots) {
                pairs.add(itemHandler to slot)
            }
        }
    }

    //    Slot    //

    override fun getSlots(): Int = pairs.size

    override fun getStackInSlot(slot: Int): ItemStack {
        val (handler: HiiragiItemHandler, index: Int) = getSlotHandler(slot)
        return handler.getStackInSlot(index)
    }

    override fun setStackInSlot(slot: Int, stack: ItemStack) {
        val (handler: HiiragiItemHandler, index: Int) = getSlotHandler(slot)
        handler.setStackInSlot(index, stack)
    }

    override fun getSlotLimit(slot: Int): Int = 64

    private fun getSlotHandler(slot: Int): Pair<HiiragiItemHandler, Int> = pairs[slot]

    //    Extraction    //

    //指定したスロットが搬出可能な場合のみアイテムを取り出すメソッド
    override fun extractItem(slot: Int, amount: Int, simulate: Boolean): ItemStack {
        return if (canExtract(slot)) {
            val (handler: HiiragiItemHandler, index: Int) = getSlotHandler(slot)
            handler.extractItem(index, amount, simulate)
        } else ItemStack.EMPTY
    }

    //アイテムを取り出せるか判定するメソッド
    private fun canExtract(slot: Int): Boolean = getSlotHandler(slot).first.ioType.canExtract

    //    Insertion    //

    //指定したスロットが搬入可能な場合のみアイテムを入れるメソッド
    override fun insertItem(slot: Int, stack: ItemStack, simulate: Boolean): ItemStack {
        return if (canInsert(slot)) {
            val (handler: HiiragiItemHandler, index: Int) = getSlotHandler(slot)
            handler.insertItem(index, stack, simulate)
        } else stack
    }

    //アイテムを入れられるか判定するメソッド
    private fun canInsert(slot: Int): Boolean = getSlotHandler(slot).first.ioType.canInsert

    //    INBTSerializable    //

    override fun serializeNBT(): NBTTagCompound {
        val nbtTagList = NBTTagList()
        for (i: Int in pairs.indices) {
            val (handler: HiiragiItemHandler, index: Int) = getSlotHandler(i)
            val stack: ItemStack = handler.getStackInSlot(index)
            if (!stack.isEmpty) NBTTagCompound().also { tag: NBTTagCompound ->
                tag.setInteger("Slot", i)
                stack.writeToNBT(tag)
                nbtTagList.appendTag(tag)
            }
        }
        return NBTTagCompound().also { it.setTag("Items", nbtTagList) }
    }

    override fun deserializeNBT(nbt: NBTTagCompound) {
        val tagList: NBTTagList = nbt.getTagList("Items", Constants.NBT.TAG_COMPOUND)
        (0 until tagList.tagCount()).forEach { i: Int ->
            val tag: NBTTagCompound = tagList.getCompoundTagAt(i)
            val slot: Int = tag.getInteger("Slot")
            if (slot >= 0 && slot < pairs.size) {
                val (handler: HiiragiItemHandler, index: Int) = getSlotHandler(i)
                handler.setStackInSlot(index, ItemStack(tag))
            }
        }
    }

    //    Custom    //

    fun isEmpty(): Boolean {
        return (0 until slots)
            .map { slot: Int -> getStackInSlot(slot) }
            .all { stack: ItemStack -> stack.isEmpty }
    }

    fun clear(): Unit = clear(0 until slots)

    fun clear(range: IntRange) {
        range.forEach { slot: Int -> setStackInSlot(slot, ItemStack.EMPTY) }
    }

}